crates/ty_ide/src/completion.rs

```diff
@@ -294,36 +310,51 @@ pub fn completion<'db>(
     completions
 }
 
-/// Adds a subset of completions derived from keywords.
+/// Adds completions derived from keywords.
 ///
-/// Note that at present, these should only be added to "scoped"
-/// completions. i.e., This will include `None`, `True`, `False`, etc.
-fn add_keyword_value_completions<'db>(
+/// This should generally only be used when offering "scoped" completions.
+/// This will include keywords corresponding to Python values (like `None`)
+/// and general language keywords (like `raise`).
+fn add_keyword_completions<'db>(
     db: &'db dyn Db,
     query: &QueryPattern,
     completions: &mut Vec<Completion<'db>>,
 ) {
-    let keywords = [
+    let keyword_values = [
         ("None", Type::none(db)),
         ("True", Type::BooleanLiteral(true)),
         ("False", Type::BooleanLiteral(false)),
     ];
-    for (name, ty) in keywords {
+    for (name, ty) in keyword_values {
         if !query.is_match_symbol_name(name) {
             continue;
         }
-        completions.push(Completion {
-            name: ast::name::Name::new(name),
-            insert: None,
-            ty: Some(ty),
-            kind: None,
-            module_name: None,
-            import: None,
-            is_type_check_only: false,
-            builtin: true,
-            documentation: None,
-        });
+        completions.push(Completion::value_keyword(name, ty));
     }
+
+    let keywords = [
+        "and", "as", "assert", "async", "await", "break", "class", "continue", "def", "del",
+        "elif", "else", "except", "finally", "for", "from", "global", "if", "import", "in", "is",
+        "lambda", "nonlocal", "not", "or", "pass", "raise", "return", "try", "while", "with",
+        "yield", "case", "match", "type",
+    ];
+    for name in keywords {
+        if !query.is_match_symbol_name(name) {
+            continue;
+        }
+        completions.push(Completion::keyword(name));
+    }
+
+    // If this assert trips, it probably means that a new keyword
+    // was added to `TokenKind` but wasn't added above. Just add it
+    // to the end of `keywords` (or `keyword_values` if the keyword
+    // is also a value expression). ---AG
+    debug_assert_eq!(
+        keyword_values.len() + keywords.len(),
+        TokenKind::iter().filter(|token| token.is_keyword()).count(),
+        "Expected actual number of keyword tokens to match \
+         number of keyword completion tokens",
+    );
```

Comment:
Hmm, I'm not sure this warants adding `strum` to `TokenKind`. 

If we go the `strum` approach, I'd iterate over all tokens above and add a `is_value_keyword` to `TokenKind` instead. 

I personally would probably just keep what you had. It seemed very reasonable to me.

Another way to avoid this is to move this to a test and assert that every `TokenKind::iter().filter(TokenKind::is_keyword)` is within the suggested completions.

Reply:
> Hmm, I'm not sure this warants adding `strum` to `TokenKind`.

yeah, I was in two minds about even making the suggestion, since it involves adding `strum` as a dependency to the parser crate 😄

If we moved this to become a test, then I suppose it could be a test-dependency only for the parser crate.

Reply:
OK, I'll just drop this commit entirely and keep what I had.